src/arch/xtensa/include/arch/atomic.h

```diff
@@ -65,6 +65,40 @@ static inline int32_t arch_atomic_sub(atomic_t *a, int32_t value)
 	return (*(volatile int32_t *)&a->value);
 }
 
+static inline int32_t arch_atomic_or(atomic_t *a, int32_t value)
+{
+	int32_t result, current;
+
+	__asm__ __volatile__(
+		"1:     l32i    %1, %2, 0\n"
+		"       wsr     %1, scompare1\n"
+		"       or     %0, %1, %3\n"
+		"       s32c1i  %0, %2, 0\n"
+		"       bne     %0, %1, 1b\n"
+		: "=&a" (result), "=&a" (current)
+		: "a" (&a->value), "a" (value)
+		: "memory");
+
+	return (*(volatile int32_t *)&a->value);
```

Comment:
We should return the old value not the new value. See https://gcc.gnu.org/onlinedocs/gcc-4.1.1/gcc/Atomic-Builtins.html. This also needs fixed in atomic_add and atomic_dec.

Reply:
@lgirdwood updated